Fixed log rotation on Windows

spacecowboy opened the pull request:
Synchronizes the operations on logs to make sure that we can safely
close the file stream before performing rotation.

Previously we did not close the stream which prevented log rotation
from ever working on Windows.

spacecowboy committed:
Log when rotation of log files failed

spacecowboy committed:
Fixed log rotation on Windows

Synchronizes the operations on logs to make sure that we can safely
close the file stream before performing rotation.

Previously we did not close the stream which prevented log rotation
from ever working on Windows.

spacecowboy committed:
Fix logically incorrect rotationThresholdExceeded method

This was actually not causing any issues because we don't create this
class when rotationThresholdBytes is 0, and we guard against the
output file not existing inside the rotation logic.

But it was still wrong.

Added a test to make sure it doesn't break while I was at it.

lutovich commented on community/logging/src/main/java/org/neo4j/logging/RotatingFileOutputStreamSupplier.java:
```diff
@@ -129,10 +130,49 @@ public RotatingFileOutputStreamSupplier( FileSystemAbstraction fileSystem, File
+            @Override
+            public void write( int i ) throws IOException
+            {
+                synchronized ( outRef )
```
Now we always synchronize on `outRef`. Do we still need it to be an `AtomicReference`?

spacecowboy replied:
Probably not, but it doesn't negatively affect anything.

lutovich replied:
Well, it makes code more complex.

lutovich commented on community/logging/src/main/java/org/neo4j/logging/RotatingFileOutputStreamSupplier.java:
```diff
@@ -190,40 +231,58 @@ private void rotate()
-            try
+        Runnable runnable = () ->
+        {
+            synchronized ( outRef )
```
Maybe we can use `ReadWriteLock` instead of synchronized, write lock here and in `#close()` and read lock in all other places?

spacecowboy replied:
We never read from the log files though, we only write to them. So it would be write locks everywhere.

lutovich replied:
`outRef` can be read-locked for everything except rotation and closing to make sure all those operations can run concurrently. It should be write-locked for rotation and close to ensure that we only rotate or close at a point in time. It does not really matter what we do with the file, some operations just can run concurrently and some can't.

spacecowboy replied:
Sure, I can buy that reasoning

spacecowboy committed:
Handle rotation failure in a better way

Previously it would just be a silent failure.

Now, if we really can't open a new file after log rotation, then fall
back to stdErr so that

* user at least has a chance to detect the problem
* we always have a valid OutputStream to log to

Also fix tests which had assumptions which no longer hold (the
OutputStream is wrapped so it never changes from the view of the user)

spacecowboy committed:
Don't log if log file can't be opened after rotation

spacecowboy committed:
Use ReadWriteLock for log rotation

Changes the locking to use ReadWriteLock instead of synchronized
blocks. This should help to increase throughput on the logging while
keeping the required exclusive locks on Closing and Rotation.

Changed OutRef to be as a simple Object instead of a AtomicReference
since we do locking and wrap the object.

Removed all references to rotated log files. There should be no need
to keep holding on to these since we close the log file before rotation.

spacecowboy committed:
Some code style fixes

spacecowboy commented:
@lutovich please have a look at the change to ReadWriteLock.

lutovich commented:
@spacecowboy looks good!

spacecowboy committed:
Make sure rotation is set to false in finally block

`shouldSurviveFilesystemErrors` had become flaky, this makes the code
more robust.
